Per table crypto documentation updates (#342)
apache/accumulo-website: _docs-2/security/on-disk-encryption.md

---

For an additional layer of security, Accumulo can encrypt files stored on disk.  On Disk encryption was reworked 

> for 2.0, making it easier to configure and more secure.  Starting with 2.1 On Disk Encryption can now be configured

Reviewer: ```suggestion
for 2.0, making it easier to configure and more secure.  Starting with 2.1, On Disk Encryption can now be configured
```

per table as well as for the entire instance (all tables). The files that can be encrypted include: [RFiles][design] and Write Ahead 
Logs (WALs). NOTE: This feature is considered experimental. For more information, see the [notes below](#things-to-keep-in-mind).

> 

Reviewer 2: I think we should drop another note here that points people to _docs-2/administration/upgrading.md. Then, in upgrading.md we should add a warning that upgrading an encrypted instance to 2.1.0 is not currently supported.

Author: I'm adding some comments about this in my latest update but is there any advice we should give to people in this case?

## Configuration

To encrypt tables on disk, encryption must be enabled before an Accumulo instance is initialized. This is

> done by configuring a crypto service factory. If on disk encryption is enabled on an existing cluster, only files

Reviewer: ```suggestion
done by configuring a crypto service factory. If on-disk encryption is enabled on an existing cluster, only files
```

Reviewer: looks like there are multiple places that could make this change

> created after it is enabled will be encrypted (root and metadata tables will not be encrypted in this case) and
> existing data won't be encrypted until compaction. 

Reviewer 3: The root and metadata tables should be encrypted once they are compacted, assuming the user configured it this way.
```suggestion
created after it is enabled will be encrypted and existing data won't be encrypted until compaction. 
```

### Encrypting All Tables

To encrypt all tables, the generic crypto service factory can be used, `GenericCryptoServiceFactory`. This factory
is useful for general purpose on disk encryption with no table context.
```
instance.crypto.opts.factory=org.apache.accumulo.core.spi.crypto.GenericCryptoServiceFactory
```

> The GenericCryptoServiceFactory requires configuring a crypto service to load and this can be done by adding the
> {% plink general.custom.crypto.service %} property to your `accumulo.properties` file.  The value of this property is the

Reviewer 3: You could just tell users to set the property, since there are multiple ways to do this.
```suggestion
The GenericCryptoServiceFactory requires configuring a crypto service to load and this can be done by setting the
{% plink general.custom.crypto.service %} property.  The value of this property is the
```

class name of the service which will perform crypto on RFiles and WALs. 
```
general.custom.crypto.service=org.apache.accumulo.core.spi.crypto.AESCryptoService
```

### Per Table Encryption

To encrypt per table, the per table crypto service factory can be used, `PerTableCryptoServiceFactory`. This factory
will load a crypto service configured by table. 
```
instance.crypto.opts.factory=org.apache.accumulo.core.spi.crypto.PerTableCryptoServiceFactory
```

> The PerTableCryptoServiceFactory requires configuring a crypto service to load for the table RFiles and this can be done by adding the

Reviewer: ```suggestion
The `PerTableCryptoServiceFactory` requires configuring a crypto service to load for the table RFiles and this can be done by adding the
```

{% plink table.crypto.opts.service %} property to a table. Example in the accumulo shell:
```

> createtable table1 -prop table.crypto.opts.service=org.apache.accumulo.core.spi.crypto.AESCryptoService

Reviewer 2: what happens if you set `table.crypto.opts.service=org.apache.accumulo.core.spi.crypto.AESCryptoService` in accumulo.properties? Are all tables encrypted then?

Reviewer 3: This would be the same as any table properties set at the site configuration (accumulo.properites), all tables would get this property. You would still need to set the Factory for it to actually work though.

Author: I haven't tried that but I would assume it would do nothing and be ignored. That property should only apply at the per table context but @milleruntime can confirm or I can test it when I get a chance to verify.

Author: Ah @milleruntime  just beat me to it, so I guess it would apply to all tables.

Reviewer 2: thx. I'm working on integrating this into the terraform configuration. I did not put this property into accumulo.properties.

```

> The PerTableCryptoServiceFactory also requires configuring a recovery and WAL crypto service by adding the following

Reviewer: ```suggestion
The `PerTableCryptoServiceFactory` also requires configuring a recovery and WAL crypto service by adding the following
```

```
For more help getting started see {% jlink org.apache.accumulo.core.security.crypto.impl.AESCryptoService %}.

## Disabling Crypto

> Crypto can be disabled by setting the property `general.custom.crypto.enabled` to false.

Reviewer 2: Does this disable for the entire instance? Is there a way to disable for a table when using the PerTableCryptoServiceFactory?

Author: Good question, this was something quick that I used just to test if I could disable it last week and have things still work and this is a section that needs to be elaborated on and isn't quite finished. I will take a look tomorrow and see where it gets applied as off the top of my head I don't remember if this only works for instance wide crypto and if there is a different property for per table to disable.

Reviewer 3: See my comment here: https://github.com/apache/accumulo/issues/2930#issuecomment-1249615992
